app/src/lib/markdown-filters/node-filter.ts

```diff
@@ -0,0 +1,88 @@
+import { EmojiFilter } from './emoji-filter'
+
+export interface INodeFilter {
+  /**
+   * Creates a document tree walker filtered to the nodes relevant to the node filter.
+   *
+   * Examples:
+   * 1) An Emoji filter operates on all text nodes, but not inside pre or code tags.
+   * 2) The issue mention filter operates on all text nodes, but not inside pre, code, or anchor tags
+   */
+  createFilterTreeWalker(doc: Document): TreeWalker
+
+  /**
+   * This filter accepts a document node and searches for it's pattern within it.
+   *
+   * If found, returns an array of nodes to replace the node with.
+   *    Example: [Node(contents before match), Node(match replacement), Node(contents after match)]
+   * If not found, returns null
+   *
+   * This is asynchronous as some filters have data must be fetched or, like in
+   * emoji, the conversion to base 64 data uri is asynchronous
+   * */
+  filter(node: Node): Promise<ReadonlyArray<Node> | null>
+}
+
+/**
+ * Builds an array of node filters to apply to markdown html. Referring to it as pipe
+ * because they will be applied in the order they are entered in the returned
+ * array. This is important as some filters impact others.
+ *
+ * @param emoji Map from the emoji shortcut (e.g., :+1:) to the image's local path.
+ */
+export function buildCustomMarkDownNodeFilterPipe(
+  emoji: Map<string, string>
+): ReadonlyArray<INodeFilter> {
+  return [new EmojiFilter(emoji)]
```

Comment:
Can we cache this? Maybe with `memoize` or something (if it works well with a `Map<string, string>`). You can even set an expiration time IIRC.

If I'm not wrong, the emojis shouldn't change while the app is running, so having to create all these all the time feels like a waste of resources if the sandboxed markdown component is recreated very frequently 😳